### DISCUSSION TITLE
I am trying to get my toplevel for file dialog

### DISCUSSION THREAD

**Discussion by pmankow** [AvaloniaUI/Avalonia#13599] [Q&A] [answered]
I am running var topLevel = TopLevel.GetTopLevel(this); in my mainviewmodel in order to open a save file dialog. However, I get this error:

CS1503	Argument 1: cannot convert from 'xxxxxxxxxx.ViewModels.MainViewModel' to 'Avalonia.Visual?'

Does anyone know how to remedy this?

**Comment by IanRawley** [comment]
"this" has to be an Avalonia Visual, and so can't be the ViewModel.  Call it in the code behind of the view, or pass a control (likely the view) to the ViewModel.

**Comment by thevortexcloud** [comment]
If you are using ReactiveUI there is a better way to do this. As I described here: https://github.com/AvaloniaUI/Avalonia/discussions/13484#discussioncomment-7478892

In any case it's considered to be bad practice to try to directly open dialogues from a view model as it does not follow the MVVM pattern properly.

**Reply by pmankow** [reply]
I'm not using Reactive UI. This is what worked:

```
// Method to get the TopLevel
private WindowBase GetTopLevel()
{
    if (Application.Current.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopLifetime)
    {
        return desktopLifetime.MainWindow;
    }

    return null;
}
```

Called by this:

`var topLevel = TopLevel.GetTopLevel(GetTopLevel());`

**Reply by DevAM-Tools** [reply]
Alternatively you might set it once at the beginning in your App:

```
public partial class App : Application
{
    public static TopLevel TopLevel { get; private set; }

    public static MainViewModel MainViewModel { get; private set; }

    public override void OnFrameworkInitializationCompleted()
    {
        MainViewModel = new();

        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopApp)
        {
            desktopApp.MainWindow = new MainWindow
            {
                DataContext = MainViewModel
            };
            TopLevel = TopLevel.GetTopLevel(desktopApp.MainWindow);
        }
        else if (ApplicationLifetime is ISingleViewApplicationLifetime singleViewPlatform)
        {
            singleViewPlatform.MainView = new MainView
            {
                DataContext = MainViewModel
            };
            TopLevel = TopLevel.GetTopLevel(singleViewPlatform.MainView);
        }

        base.OnFrameworkInitializationCompleted();
    }
}
```

Then you can access it anywhere it is needed with `App.TopLevel`.

**Reply by NickBarton** [reply]
Couldn't make this work Avalonia 11.0.10.
GetTopLevel returns null.

**Reply by thevortexcloud** [reply]
You should only ever try to access a top level from the code behind of your views. Under MVVM, a view model has nothing to do with a view or top level.

**Reply by DevAM-Tools** [reply]
> Couldn't make this work Avalonia 11.0.10.

I don't know for sure, but `TopLevel.GetTopLevel()` might only produce useful results in a desktop app. (I am using Avalonia 11.x on Windows)

> You should only ever try to access a top level from the code behind of your views. Under MVVM, a view model has nothing to do with a view or top level.

I disagree. `TopLevel?.StorageProvider?.OpenFilePickerAsync(openOptions)` or `TopLevel?.Clipboard()` might be something that you want to use in your ViewModel.

**Reply by thevortexcloud** [reply]
You may disagree, but it's still an MVVM violation. The viewmodel is not allowed to know about/directly access anything on the view. There are ways (which I linked in my original comment) to access things like a file picker indirectly from a view model which will not violate MVVM rules.

**Comment by timunie** [comment, marked as answer]
I have a sample in preview. Unfortunately I don't find the time to finish it due to life throwing in. If you want to check it out, any comment on the samples would be welcome:

https://github.com/timunie/Avalonia.Samples/tree/feature/MVVM_Dialogs/src/Avalonia.Samples/ViewInteraction 

Thanks
